### PR TITLE
[llvm-objcopy] Add change-section-lma *+/-offset

### PR DESCRIPTION
**Pull request by eleanor-arm**
llvm-objcopy did not support change-section-lma argument.

This patch adds support for a use case of change-section-lma, that is
shifting load address of all sections by the same offset. This seems to
be the only practical use case of change-section-lma, found in other
software such as Zephyr RTOS's build system.

This is an option that could possibly be supported in some other than
ELF formats, however this change only implements it for ELF. When used
with other formats an error message is raised.

In comparison, the behavior of GNU objcopy is inconsistent. For some ELF
files it behaves the same as described above. For others, it copies the
file without modifying the p_paddr fields when it would be expected. In
some experiments it modifies arbitrary fields in section or program
headers. It is unclear what exactly determines this.
The executable file generated by yaml2obj in this test is not parsable
by GNU objcopy. With Machine set to EM_AARCH64, the file can be parsed
and the first test in the test file completes with 0 exit code. However,
the result is rather arbitrary. AArch64 GNU objcopy subtracts 0x1000
from p_filesz and p_memsz of the first LOAD section and 0x1000 from
p_offset of the second LOAD section. It does not look meaningful.

### CODE REVIEW

File: llvm/tools/llvm-objcopy/ObjcopyOptions.cpp [hunk 1]
```diff
@@ -552,6 +552,33 @@ static Error loadNewSectionData(StringRef ArgValue, StringRef OptionName,
...
+                                               StringRef OptionName) {
+  StringRef StringValue;
+  if (ArgValue.starts_with("*+")) {
+    StringValue = ArgValue.slice(2, StringRef::npos);
+  } else if (ArgValue.starts_with("*-")) {
+    StringValue = ArgValue.slice(1, StringRef::npos);
+  } else {
+    return createStringError(errc::invalid_argument,
+                             "bad format for " + OptionName +
+                                 ": it is required that all sections "
```
Comment (jh7370) [inline llvm/tools/llvm-objcopy/ObjcopyOptions.cpp:565]: I think you should rephrase this to something like "changing a specific section LMA is not supported. Use *+val or *-val instead"

File: llvm/tools/llvm-objcopy/ObjcopyOptions.cpp [hunk 2]
```diff
@@ -552,6 +552,33 @@ static Error loadNewSectionData(StringRef ArgValue, StringRef OptionName,
...
+                                 "are either incremented, or decremented at "
+                                 "the same time; use *+val, "
+                                 "or *-val");
+  }
+  if (StringValue.empty())
+    return createStringError(errc::invalid_argument,
+                             "bad format for " + OptionName +
+                                 ": missing offset of LMA");
```
Comment (jh7370) [inline llvm/tools/llvm-objcopy/ObjcopyOptions.cpp:573]:
```suggestion
                                 ": missing LMA offset");
```
would read slightly better

File: llvm/tools/llvm-objcopy/ObjcopyOptions.cpp [hunk 3]
```diff
@@ -552,6 +552,33 @@ static Error loadNewSectionData(StringRef ArgValue, StringRef OptionName,
...
+
+  auto SLMAV = getAsInteger<int64_t>(StringValue);
+  if (!SLMAV)
+    return createStringError(SLMAV.getError(),
+                             "Unable to parse adjustment value");
```
Comment (jh7370) [inline llvm/tools/llvm-objcopy/ObjcopyOptions.cpp:578]: Please see the coding standards for how to format error messages.

File: llvm/tools/llvm-objcopy/ObjcopyOptions.cpp [hunk 4]
```diff
@@ -552,6 +552,33 @@ static Error loadNewSectionData(StringRef ArgValue, StringRef OptionName,
...
+                                 "are either incremented, or decremented at "
+                                 "the same time; use *+val, "
+                                 "or *-val");
+  }
+  if (StringValue.empty())
+    return createStringError(errc::invalid_argument,
+                             "bad format for " + OptionName +
+                                 ": missing offset of LMA");
+
+  auto SLMAV = getAsInteger<int64_t>(StringValue);
```
Comment (jh7370) [inline llvm/tools/llvm-objcopy/ObjcopyOptions.cpp:575]: I have no idea what `SLMAV` is supposed to mean. Please don't abbreviate variable names unnecessarily.

File: llvm/test/tools/llvm-objcopy/ELF/change-section-lma-single-load.test [hunk 5]
```diff
@@ -0,0 +1,45 @@
+# RUN: yaml2obj %s -o %t
```
Comment (jh7370) [inline llvm/test/tools/llvm-objcopy/ELF/change-section-lma-single-load.test:1]: It would be preferable to have all test cases in a single file. You can use yaml2obj's --docnum option to have multiple different YAML inputs in the same file (see other tests for examples).

That being said, what's the benefit of the "single-load" test case that the "multiple-load" case doesn't provide?

File: llvm/test/tools/llvm-objcopy/ELF/change-section-lma-multiple-load.test [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+# RUN: llvm-readelf -l %t3 | FileCheck %s --check-prefix=CHECK-MINUS
+# RUN: not llvm-objcopy --change-section-lma .text3=0x5000 %t  2>&1 | FileCheck %s --check-prefix=ERR
+
+!ELF
+FileHeader:
+  Class:           ELFCLASS64
+  Data:            ELFDATA2LSB
+  Type:            ET_EXEC
+  Machine:         EM_AARCH64
+Sections:
```
Comment (jh7370) [inline llvm/test/tools/llvm-objcopy/ELF/change-section-lma-multiple-load.test:14]: If I'm not mistaken, I think the sections are actually irrelevant to the test case and code behaviour (you can explicitly set the p_filesz of the segment in the YAML, without needing sections to drive it). If so, let's omit them.

File: llvm/test/tools/llvm-objcopy/ELF/change-section-lma-multiple-load.test [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+# RUN: llvm-objcopy --change-section-lma *-0x10 %t2 %t3
+# RUN: llvm-readelf -l %t3 | FileCheck %s --check-prefix=CHECK-MINUS
+# RUN: not llvm-objcopy --change-section-lma .text3=0x5000 %t  2>&1 | FileCheck %s --check-prefix=ERR
+
+!ELF
+FileHeader:
+  Class:           ELFCLASS64
+  Data:            ELFDATA2LSB
+  Type:            ET_EXEC
+  Machine:         EM_AARCH64
```
Comment (jh7370) [inline llvm/test/tools/llvm-objcopy/ELF/change-section-lma-multiple-load.test:13]: The `Machine` value is irrelevant for the test case, so you can omit it.

File: llvm/test/tools/llvm-objcopy/ELF/change-section-lma-multiple-load.test [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+  - Name:            .text3
+    Type:            SHT_PROGBITS
+    Flags:           [ SHF_ALLOC, SHF_EXECINSTR ]
+    Address:         0x3000
+    AddressAlign:    0x1000
+    Content:         "c3c3c3c3"
+    Size:            0x1000
+ProgramHeaders:
+  - Type:     PT_LOAD
+    Flags:    [ PF_R ]
```
Comment (jh7370) [inline llvm/test/tools/llvm-objcopy/ELF/change-section-lma-multiple-load.test:38]: I believe the `Flags` are irrelevant, so please omit them, to reduce the noise in the test.

File: llvm/test/tools/llvm-objcopy/ELF/change-section-lma-multiple-load.test [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    VAddr:    0x1000
+    FirstSec: .text1
+    LastSec:  .text2
+# CHECK-PLUS:  Type           Offset   VirtAddr           PhysAddr           FileSiz  MemSiz
```
Comment (jh7370) [inline llvm/test/tools/llvm-objcopy/ELF/change-section-lma-multiple-load.test:42]: Once you've simplified the YAML as suggested, it would be more readable to order the test something like this, I think:
```
# RUN: ...
# RUN: ...

# CHECK...
# CHECK...

<yaml>
```

File: llvm/test/tools/llvm-objcopy/ELF/change-section-lma-multiple-load.test [hunk 10]
```diff
@@ -0,0 +1,52 @@
+# RUN: yaml2obj %s -o %t
```
Comment (jh7370) [inline llvm/test/tools/llvm-objcopy/ELF/change-section-lma-multiple-load.test:1]: I think you additionally need:
1) A segment with 0 file size.
2) A segment with an explicit p_paddr that is different to the p_vaddr.
3) A non PT_LOAD segment (with non-zero p_filesz), to show that it's not just PT_LOAD segments that are impacted (should they be impacted though? what does GNU objcopy do?).
4) A nested segment (with non-zero p_filesz) to show that segments within other segments are appropriately updated. The segment should probably have a non-zero offset within the PT_LOAD it's nested inside.
5) A segment with 0 p_vaddr/p_paddr.
6) Test cases that cover all the different new error messages you've added for parsing the option.

File: llvm/include/llvm/ObjCopy/CommonConfig.h [hunk 11]
```diff
@@ -245,6 +245,9 @@ struct CommonConfig {
   // Symbol info specified by --add-symbol option.
   SmallVector<NewSymbolInfo, 0> SymbolsToAdd;
 
+  // Integer options
+  int64_t ChangeSectionLMAValAll = 0;
```
Comment (jh7370) [inline llvm/include/llvm/ObjCopy/CommonConfig.h:249]: Should this option really be in `CommonConfig.h` and not the ELF-specific version? Is it actually applicable to non-ELF formats?

File: llvm/tools/llvm-objcopy/ObjcopyOpts.td [hunk 12]
```diff
@@ -254,6 +254,14 @@ def adjust_start : JoinedOrSeparate<["--"], "adjust-start">,
                    Alias<change_start>,
                    HelpText<"Alias for --change-start">;
 
+defm adjust_section_lma
+    : Eq<"change-section-lma",
+         "Change LMA of all sections by <val>. LMA is the physical address where the section is "
+         "loaded in the memory; as opoosed to VMA, which is the virtual address at which the "
+         "section resides during program execution. Ussually LMA and VMA are the same, but some "
+         "platforms support these to be different resulting in runtime copying">,
```
Comment (jh7370) [inline llvm/tools/llvm-objcopy/ObjcopyOpts.td:262]: This description seems too verbose for the help text, and should be in the documentation instead, if needed at all.

File: llvm/tools/llvm-objcopy/ObjcopyOpts.td [hunk 13]
```diff
@@ -254,6 +254,14 @@ def adjust_start : JoinedOrSeparate<["--"], "adjust-start">,
...
+      MetaVarName<"*{+|-}val">;
```
Comment (jh7370) [inline llvm/tools/llvm-objcopy/ObjcopyOpts.td:263]: How does this appear in the --help output?

**Review by jh7370 — COMMENTED** [review]
Could you describe what the actual behaviour is when GNU objcopy uses this option, please?

Also, could you go into more detail about what your use-case is, please?

File: llvm/docs/CommandGuide/llvm-objcopy.rst [hunk 14]
```diff
@@ -528,6 +528,10 @@ them.
 
  Mark all defined global symbols as weak in the output.
 
+.. option:: --change-section-lma \*{+-}<offset>
+
+ Currently only supports changing LMA for all sections by the same offset.
```
[the remark this answers is not in the document]
Comment (eleanor-arm) [reply llvm/docs/CommandGuide/llvm-objcopy.rst]: Done

File: llvm/lib/ObjCopy/ELF/ELFObjcopy.cpp [hunk 15]
```diff
@@ -670,6 +670,13 @@ static Error handleArgs(const CommonConfig &Config, const ELFConfig &ELFConfig,
     }
   }
 
+  if (Config.ChangeSectionLMAValAll != 0) {
+    for (auto &Seg : Obj.segments()) {
```
[the remark this answers is not in the document]
Comment (eleanor-arm) [reply llvm/lib/ObjCopy/ELF/ELFObjcopy.cpp]: Fixed

File: llvm/tools/llvm-objcopy/ObjcopyOptions.cpp (hunk 1 above)
Comment (eleanor-arm) [reply llvm/tools/llvm-objcopy/ObjcopyOptions.cpp]: Fixed

File: llvm/tools/llvm-objcopy/ObjcopyOptions.cpp (hunk 2 above)
Comment (eleanor-arm) [reply llvm/tools/llvm-objcopy/ObjcopyOptions.cpp]: Fixed

File: llvm/tools/llvm-objcopy/ObjcopyOptions.cpp (hunk 3 above)
Comment (eleanor-arm) [reply llvm/tools/llvm-objcopy/ObjcopyOptions.cpp]: Fixed

File: llvm/tools/llvm-objcopy/ObjcopyOptions.cpp (hunk 4 above)
Comment (eleanor-arm) [reply llvm/tools/llvm-objcopy/ObjcopyOptions.cpp]: Done

File: llvm/test/tools/llvm-objcopy/ELF/change-section-lma-single-load.test (hunk 5 above)
Comment (eleanor-arm) [reply llvm/test/tools/llvm-objcopy/ELF/change-section-lma-single-load.test]: One yaml file seems sufficient. The fixup has only one.

File: llvm/test/tools/llvm-objcopy/ELF/change-section-lma-multiple-load.test (hunk 6 above)
Comment (eleanor-arm) [reply llvm/test/tools/llvm-objcopy/ELF/change-section-lma-multiple-load.test]: When I removed the sections the executable generated by yaml2obj was not usable by llvm-objcopy, although it looked quite similar. 

` error: program header with offset 0x1000 and file size 0x2000 goes past the end of the file`

File: llvm/test/tools/llvm-objcopy/ELF/change-section-lma-multiple-load.test (hunk 7 above)
Comment (eleanor-arm) [reply llvm/test/tools/llvm-objcopy/ELF/change-section-lma-multiple-load.test]: Done

File: llvm/test/tools/llvm-objcopy/ELF/change-section-lma-multiple-load.test (hunk 8 above)
Comment (eleanor-arm) [reply llvm/test/tools/llvm-objcopy/ELF/change-section-lma-multiple-load.test]: Done

File: llvm/test/tools/llvm-objcopy/ELF/change-section-lma-multiple-load.test (hunk 9 above)
Comment (eleanor-arm) [reply llvm/test/tools/llvm-objcopy/ELF/change-section-lma-multiple-load.test]: Interestingly, I was told to add the checks inline when contributing to other parts of LLVM but I don't mind them being all on top.

File: llvm/test/tools/llvm-objcopy/ELF/change-section-lma-multiple-load.test (hunk 10 above)
Comment (eleanor-arm) [reply llvm/test/tools/llvm-objcopy/ELF/change-section-lma-multiple-load.test]: In change-section-lma.test: 
1. PT_NOTE
2. PT_LOAD (.text2) 
3. PT_PHDR
4. PT_PHDR in PR_LOAD (.text1) 
5. PT_NOTE
6. Covers all of them but the use with unsupported file formats which seems excessive

Comment (eleanor-arm) [reply llvm/test/tools/llvm-objcopy/ELF/change-section-lma-multiple-load.test]: > The segment should probably have a non-zero offset within the PT_LOAD it's nested inside.
I am not sure how that helps with coverage.

File: llvm/include/llvm/ObjCopy/CommonConfig.h (hunk 11 above)
Comment (eleanor-arm) [reply llvm/include/llvm/ObjCopy/CommonConfig.h:249]: Possibly applicable but not implemented by this patch. I left it there for now, and added more explanation in the commit message and added the option to ConfigManager.cpp

File: llvm/tools/llvm-objcopy/ObjcopyOpts.td (hunk 12 above)
Comment (eleanor-arm) [reply llvm/tools/llvm-objcopy/ObjcopyOpts.td]: I've made the description simpler.

File: llvm/tools/llvm-objcopy/ObjcopyOpts.td (hunk 13 above)
Comment (eleanor-arm) [reply llvm/tools/llvm-objcopy/ObjcopyOpts.td:259]: It looks as expected.

File: llvm/tools/llvm-objcopy/ObjcopyOptions.cpp [hunk 16]
```diff
@@ -552,6 +552,38 @@ static Error loadNewSectionData(StringRef ArgValue, StringRef OptionName,
...
+    return createStringError(errc::invalid_argument,
+                             "bad format for " + OptionName +
+                                 ": missing LMA offset");
+
+  auto LMAValue = getAsInteger<int64_t>(StringValue);
+  if (!LMAValue)
+    return createStringError(LMAValue.getError(),
+                             "bad format for " + OptionName + ": value after " +
+                                 ArgValue.slice(0, 2) + " is " + StringValue +
+                                 " when it should be integer");
```
[the remark this answers is not in the document]
Comment (eleanor-arm) [reply llvm/tools/llvm-objcopy/ObjcopyOptions.cpp:583]: This is intentional. "integer" is meant as an adjective here as the value is obviously a string. It just needs to be convertible to an integer.

File: llvm/docs/CommandGuide/llvm-objcopy.rst [hunk 17]
```diff
@@ -528,6 +528,10 @@ them.
 
  Mark all defined global symbols as weak in the output.
 
+.. option:: --change-section-lma \*{+-}<val>
+
+ Shift LMA of non-zero-sized sections in the program header by ``<val>``
```
[the remark this answers is not in the document]
Comment (eleanor-arm) [reply llvm/docs/CommandGuide/llvm-objcopy.rst]: Fixed

File: llvm/docs/CommandGuide/llvm-objcopy.rst [hunk 18]
```diff
@@ -528,6 +528,10 @@ them.
 
  Mark all defined global symbols as weak in the output.
 
+.. option:: --change-section-lma \*{+-}<val>
```
[the remark this answers is not in the document]
Comment (eleanor-arm) [reply llvm/docs/CommandGuide/llvm-objcopy.rst]: Fixed

File: llvm/test/tools/llvm-objcopy/ELF/change-section-lma.test [hunk 19]
```diff
@@ -0,0 +1,55 @@
+# RUN: yaml2obj %s -o %t
+# RUN: llvm-objcopy --change-section-lma *+0x20 %t %t2
+# RUN: llvm-readelf -l %t2 | FileCheck %s --check-prefix=CHECK-PLUS
+# RUN: llvm-objcopy --change-section-lma *-0x10 %t2 %t3
```
[the remark this answers is not in the document]
Comment (eleanor-arm) [reply llvm/test/tools/llvm-objcopy/ELF/change-section-lma.test]: Changed, thanks for the suggestion

File: llvm/lib/ObjCopy/ELF/ELFObjcopy.cpp [hunk 20]
```diff
@@ -670,6 +670,13 @@ static Error handleArgs(const CommonConfig &Config, const ELFConfig &ELFConfig,
     }
   }
 
+  if (Config.ChangeSectionLMAValAll != 0) {
+    for (Segment &Seg : Obj.segments()) {
+      if (Seg.FileSize > 0)
+        Seg.PAddr += Config.ChangeSectionLMAValAll;
```
[the remark this answers is not in the document]
Comment (eleanor-arm) [reply llvm/lib/ObjCopy/ELF/ELFObjcopy.cpp]: That's a very good point. I've added code to check for overflow/underflow.

File: llvm/lib/ObjCopy/ELF/ELFObjcopy.cpp [hunk 21]
```diff
@@ -670,6 +670,27 @@ static Error handleArgs(const CommonConfig &Config, const ELFConfig &ELFConfig,
...
   }
 
+  if (Config.ChangeSectionLMAValAll != 0) {
+    for (Segment &Seg : Obj.segments()) {
+      if (Seg.FileSize > 0) {
+        if (Config.ChangeSectionLMAValAll > 0 &&
+            Seg.PAddr > std::numeric_limits<uint64_t>::max() -
+                            Config.ChangeSectionLMAValAll) {
+          return createStringError(errc::invalid_argument,
+                                   "address 0x" + Twine::utohexstr(Seg.PAddr) +
```
Comment (jh7370) [inline llvm/lib/ObjCopy/ELF/ELFObjcopy.cpp:680]: For this and the similar error below, I think it would be useful to give a little more context to the error, something like "would overflow when adjusting by <value>".

File: llvm/lib/ObjCopy/ELF/ELFObjcopy.cpp [hunk 22]
```diff
@@ -670,6 +670,27 @@ static Error handleArgs(const CommonConfig &Config, const ELFConfig &ELFConfig,
...
+                                       " would overflow");
+        } else if (Config.ChangeSectionLMAValAll < 0 &&
+                   Seg.PAddr < std::numeric_limits<uint64_t>::min() -
+                                   Config.ChangeSectionLMAValAll)
+          return createStringError(errc::invalid_argument,
+                                   "address 0x" + Twine::utohexstr(Seg.PAddr) +
+                                       " would underflow");
+        else
```
Comment (jh7370) [inline llvm/lib/ObjCopy/ELF/ELFObjcopy.cpp:688]: I believe LLVM coding standards direct that this `else` should have braces, since the other parts of the if/else chain do.

File: llvm/test/tools/llvm-objcopy/ELF/change-section-lma-multiple-load.test (hunk 6 above)
Comment (jh7370) [reply llvm/test/tools/llvm-objcopy/ELF/change-section-lma-multiple-load.test]: Okay, fair enough. I've actually changed my mind about not having any sections anyway, as I feel like it would be good to verify that the section sh_addr value isn't changed, as part of "documenting what the feature does through tests".

File: llvm/test/tools/llvm-objcopy/ELF/change-section-lma-multiple-load.test (hunk 10 above)
Comment (jh7370) [reply llvm/test/tools/llvm-objcopy/ELF/change-section-lma-multiple-load.test]: I honestly can't remember what my reasoning was with that case 🤷‍♂️

Comment (jh7370) [reply llvm/test/tools/llvm-objcopy/ELF/change-section-lma-multiple-load.test]: With point 3, I was explicitly referring to a segment not nested within another segment. This is important, because some tools treat PT_LOAD segments (and segments nested within them) as different to other top-level segment types, hence why I'd ideally like this checked against GNU's behaviour.

File: llvm/docs/CommandGuide/llvm-objcopy.rst [hunk 23]
```diff
@@ -299,6 +299,10 @@ them.
  Allow :program:`llvm-objcopy` to remove sections even if it would leave invalid
  section references. Any invalid sh_link fields will be set to zero.
 
+.. option:: --change-section-lma \*{+-}<val>
+
+ Shift LMA of non-zero-sized sections in the program header table by ``<val>``
```
Comment (jh7370) [inline llvm/docs/CommandGuide/llvm-objcopy.rst:304]:
```suggestion
 Shift LMA of non-zero-sized segments by ``<val>``
```
How about this? Sections aren't in the program header table, only program headers (aka segments, which are in turn implicitly in the program header table).

File: llvm/tools/llvm-objcopy/ObjcopyOptions.cpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (jh7370) [reply llvm/tools/llvm-objcopy/ObjcopyOptions.cpp:583]: "integer" isn't an adjective though, it's a noun. You can use "integral", I guess, but I think the meaning is clear enough and more grammatically correct with my suggestion.

[the remark this answers is not in the document]
Comment (eleanor-arm) [reply llvm/tools/llvm-objcopy/ObjcopyOptions.cpp]: OK. I've changed it to the original suggestion.

File: llvm/lib/ObjCopy/ELF/ELFObjcopy.cpp (hunk 21 above)
Comment (eleanor-arm) [reply llvm/lib/ObjCopy/ELF/ELFObjcopy.cpp]: Improved this as well.

File: llvm/lib/ObjCopy/ELF/ELFObjcopy.cpp (hunk 22 above)
Comment (eleanor-arm) [reply llvm/lib/ObjCopy/ELF/ELFObjcopy.cpp]: The coding standard actually discourages use of `else` after return, so I've removed it. And I've added braces around the previous `else if` because the control flow is "deeply nested".

File: llvm/docs/CommandGuide/llvm-objcopy.rst (hunk 23 above)
Comment (eleanor-arm) [reply llvm/docs/CommandGuide/llvm-objcopy.rst]: Added this suggestion to **Fixup3**

**Comment by eleanor-arm** [comment]
>  it would be good to verify that the section sh_addr value isn't changed, as part of "documenting what the feature does through tests".

I've tried to add this to the test. Is it clear enough what it's showing?

File: llvm/test/tools/llvm-objcopy/ELF/change-section-lma.test [hunk 24]
```diff
@@ -1,32 +1,45 @@
 # RUN: yaml2obj %s -o %t
 # RUN: llvm-objcopy --change-section-lma *+0x20 %t %t2
-# RUN: llvm-readelf -l %t2 | FileCheck %s --check-prefix=CHECK-PLUS
+# RUN: llvm-readelf --program-headers %t2 | FileCheck %s --check-prefix=CHECK-PLUS-PROGRAMS
+# RUN: llvm-readelf --section-headers %t2 | FileCheck %s --check-prefix=CHECK-PLUS-SECTIONS
```
Comment (jh7370) [inline llvm/test/tools/llvm-objcopy/ELF/change-section-lma.test:4]: You could simplify this and the similar below down to a single call with both options specified.

**Review by jh7370 — COMMENTED** [review]
This is mostly fine, but it looks like my last round of test comments haven't been addressed yet?

File: llvm/test/tools/llvm-objcopy/ELF/change-section-lma.test (hunk 24 above)
Comment (eleanor-arm) [reply llvm/test/tools/llvm-objcopy/ELF/change-section-lma.test:4]: Yes the original version had them in one command but I "simplified" it by having two separate commands. I think it makes it clearly to see what is being checked.

File: llvm/test/tools/llvm-objcopy/ELF/change-section-lma-multiple-load.test (hunk 10 above)
Comment (eleanor-arm) [reply llvm/test/tools/llvm-objcopy/ELF/change-section-lma-multiple-load.test]: I see. For point 3, I will add another non-zero-size NOTE header in fixup 4 that is independent from the LOAD headers.
As mentioned before in other places I don't see a meaning in GNU's objcopy actual behaviour with this particular option. However, I defined what the LLVM version will do based on GNU documentation, some bits and pieces of GNU behaviour and how I saw the option being used in real software. 

The zero-size NOTE is already independent as it automatically gets virtual address of 0x0.

File: llvm/test/tools/llvm-objcopy/ELF/change-section-lma-multiple-load.test (hunk 6 above)
Comment (eleanor-arm) [reply llvm/test/tools/llvm-objcopy/ELF/change-section-lma-multiple-load.test]: Hopefully the latest version of the test shows this.

**Comment by eleanor-arm** [comment]
I've responded to all the unresolved comments. Hopefully, this should cover it all.

File: llvm/test/tools/llvm-objcopy/ELF/change-section-lma-multiple-load.test (hunk 10 above)
Comment (jh7370) [reply llvm/test/tools/llvm-objcopy/ELF/change-section-lma-multiple-load.test]: The non-empty PT_NOTE is currently nested inside the first PT_LOAD segment, if I'm not mistaken? I think it needs to have a companion section that can be assigned to it, to force the offset to be not within the other PT_LOADs.

> The zero-size NOTE is already independent as it automatically gets virtual address of 0x0.

Not quite what I meant (I wanted a segment with 0 virtual address, and a different segment with zero file size). That being said, I actually don't think we need a 0 virtual address case, having given it a bit more thought, since there's nothing special about this case.

File: llvm/docs/CommandGuide/llvm-objcopy.rst [hunk 25]
```diff
@@ -299,6 +299,10 @@ them.
  Allow :program:`llvm-objcopy` to remove sections even if it would leave invalid
  section references. Any invalid sh_link fields will be set to zero.
 
+.. option:: --change-section-lma \*{+-}<val>
+
+ Shift LMA of non-zero-sized segments by ``<val>``
```
Comment (jh7370) [inline llvm/docs/CommandGuide/llvm-objcopy.rst:304]:
```suggestion
 Shift LMA of non-zero-sized segments by ``<val>``.
```

**Review by jh7370 — APPROVED** [review]
One nit, otherwise LGTM.

File: llvm/test/tools/llvm-objcopy/ELF/change-section-lma-multiple-load.test (hunk 10 above)
Comment (eleanor-arm) [reply llvm/test/tools/llvm-objcopy/ELF/change-section-lma-multiple-load.test]: Ah OK, I've changed the test so that the non-zero PT_NOTE section is elsewhere in the file. And then I thought it was better to make the 0-size segment independent too (it might not really matter considering it is zero size).
